aips/aip-20.md

```diff
@@ -0,0 +1,464 @@
+---
+aip: 20
+title: Generic Cryptography Algebra and BLS12-381 Implementation
+author: zhoujun-ma, alin
+discussions-to (*optional): https://github.com/aptos-foundation/AIPs/issues/94
+Status: Draft
+type: Standard (Framework)
+created: 2023/03/15
+---
+
+# AIP-14 - Generic Cryptography Algebra and BLS12-381 Implementation
+## Summary
+
+This AIP proposes the support of generic cryptography algebra operations in Aptos standard library.
+
+The initial list of the supported generic operations includes group/field element serialization/deserialization, basic arithmetic, pairing, hash-to-structure, casting.
+
+The initial list of supported algebraic structures includes groups/fields used in BLS12-381, a popular pairing-friendly curve as described [here](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-11#name-bls-curves-for-the-128-bit-).
+
+Either the operation list or the structure list can be extended by future AIPs.
+
+## Motivation
+
+Algebraic structures are fundamental building blocks for many cryptographic schemes, but also hard to implement efficiently in pure Move.
+This change should allow Move developers to implement generic constructions of those schemes, then get different instantiations by only switching the type parameter(s).
+
+For example, if BLS12-381 groups and BN254 groups are supported, one can implement a generic Groth16 proof verifier construction, then be able to use both BLS12-381-based Groth16 proof verifier and BN254-based Groth16 proof verifier.
+
+BLS12-381-based Groth16 proof verifier has been implemented this way as part of the reference implementation.
+
+## Rationale
+
+An alternative non-generic approach is to expose instantiated schemes directly in stdlib.
+For example, we can define a Groth16 proof verification function
+`0x1::groth16_b::verify_proof(vk, proof, public_inputs): bool`
+for every bilinear map `b`;
```

Comment:
Editorial suggestion: replace ';' by a dot. Move "for ECDSA" as a "For ECDSA" new paragraph.

Reply:
Editorial suggestion #2: Replace `b` by `<curve>` and say "for every pairing-friendly elliptic curve <curve>"

Reply:
Or maybe `<group>`